File: app/models/notebook.py
import uuid
from datetime import datetime
from .db import db, environment, SCHEMA


class Note(db.Model):
    __tablename__ = "notes"
    if environment == "prod":
        __table_args__ = {"schema": SCHEMA}

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = db.Column(
        db.UUID(as_uuid=True), db.ForeignKey("users.id"), nullable=False
    )
    title = db.Column(db.String(255), nullable=False)
    content = db.Column(db.Text, nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    updated_at = db.Column(
        db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )
    tags = db.Column(db.String(255))  # Comma-separated tags

    user = db.relationship("User", back_populates="notes")

    def to_dict(self):
        return {
            "id": str(self.id),
            "user_id": str(self.user_id),
            "title": self.title,
            "content": self.content,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
            "tags": self.tags.split(",") if self.tags else [],
        }


class Experiment(db.Model):
    __tablename__ = "experiments"
    if environment == "prod":
        __table_args__ = {"schema": SCHEMA}

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = db.Column(
        db.UUID(as_uuid=True), db.ForeignKey("users.id"), nullable=False
    )
    title = db.Column(db.String(255), nullable=False)
    hypothesis = db.Column(db.Text, nullable=False)
    materials = db.Column(db.Text)
    methods = db.Column(db.Text, nullable=False)
    results = db.Column(db.Text)
    conclusion = db.Column(db.Text)
    status = db.Column(
        db.String(50), default="planned"
    )  # planned, in_progress, completed, failed
    started_at = db.Column(db.DateTime)
    completed_at = db.Column(db.DateTime)
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    updated_at = db.Column(
        db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )
    references = db.Column(db.Text)

    user = db.relationship("User", back_populates="experiments")
    steps = db.relationship(
        "ExperimentStep", back_populates="experiment", cascade="all, delete-orphan"
    )
    attachments = db.relationship(
        "ExperimentAttachment",
        back_populates="experiment",
        cascade="all, delete-orphan",
    )

    def to_dict(self):
        return {
            "id": str(self.id),
            "user_id": str(self.user_id),
            "title": self.title,
            "hypothesis": self.hypothesis,
            "materials": self.materials,
            "methods": self.methods,
            "results": self.results,
            "conclusion": self.conclusion,
            "status": self.status,
            "started_at": self.started_at.isoformat() if self.started_at else None,
            "completed_at": (
                self.completed_at.isoformat() if self.completed_at else None
            ),
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
            "references": self.references,
            "steps": [step.to_dict() for step in self.steps],
            "attachments": [attachment.to_dict() for attachment in self.attachments],
        }


class ExperimentStep(db.Model):
    __tablename__ = "experiment_steps"
    if environment == "prod":
        __table_args__ = {"schema": SCHEMA}

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    experiment_id = db.Column(
        db.UUID(as_uuid=True), db.ForeignKey("experiments.id"), nullable=False
    )
    step_number = db.Column(db.Integer, nullable=False)
    description = db.Column(db.Text, nullable=False)
    observation = db.Column(db.Text)
    started_at = db.Column(db.DateTime)
    completed_at = db.Column(db.DateTime)
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    updated_at = db.Column(
        db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    experiment = db.relationship("Experiment", back_populates="steps")

    def to_dict(self):
        return {
            "id": str(self.id),
            "experiment_id": str(self.experiment_id),
            "step_number": self.step_number,
            "description": self.description,
            "observation": self.observation,
            "started_at": self.started_at.isoformat() if self.started_at else None,
            "completed_at": (
                self.completed_at.isoformat() if self.completed_at else None
            ),
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }


class ExperimentAttachment(db.Model):
    __tablename__ = "experiment_attachments"
    if environment == "prod":
        __table_args__ = {"schema": SCHEMA}

    id = db.Column(db.UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    experiment_id = db.Column(
        db.UUID(as_uuid=True), db.ForeignKey("experiments.id"), nullable=False
    )
    file_name = db.Column(db.String(255), nullable=False)
    file_type = db.Column(db.String(100), nullable=False)
    file_path = db.Column(db.String(500), nullable=False)
    description = db.Column(db.Text)
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)

    experiment = db.relationship("Experiment", back_populates="attachments")

    def to_dict(self):
        return {
            "id": str(self.id),
            "experiment_id": str(self.experiment_id),
            "file_name": self.file_name,
            "file_type": self.file_type,
            "file_path": self.file_path,
            "description": self.description,
            "created_at": self.created_at.isoformat(),
        }

File: frontend/src/services/notebook.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { useState } from 'react';

export interface Note {
  id: string;
  user_id: string;
  title: string;
  content: string;
  created_at: string;
  updated_at: string;
  tags: string[];
}

export interface ExperimentStep {
  id: string;
  experiment_id: string;
  step_number: number;
  description: string;
  observation: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExperimentAttachment {
  id: string;
  experiment_id: string;
  file_name: string;
  file_type: string;
  file_path: string;
  description: string | null;
  created_at: string;
}

export interface Experiment {
  id: string;
  user_id: string;
  title: string;
  hypothesis: string;
  materials: string | null;
  methods: string;
  results: string | null;
  conclusion: string | null;
  status: 'planned' | 'in_progress' | 'completed' | 'failed';
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
  references: string | null;
  steps: ExperimentStep[];
  attachments: ExperimentAttachment[];
}

export type CreateNoteDto = {
  title: string;
  content: string;
  tags?: string[];
};

export type UpdateNoteDto = Partial<CreateNoteDto>;

export type CreateExperimentDto = {
  title: string;
  hypothesis: string;
  materials?: string;
  methods: string;
  references?: string;
  steps?: Array<{
    description: string;
    observation?: string;
  }>;
};

export type UpdateExperimentDto = Partial<{
  title: string;
  hypothesis: string;
  materials: string;
  methods: string;
  results: string;
  conclusion: string;
  status: 'planned' | 'in_progress' | 'completed' | 'failed';
  references: string;
  steps: Array<{
    id?: string;
    description: string;
    observation?: string;
    started_at?: string;
    completed_at?: string;
  }>;
}>;

const API_BASE_URL = '/api';

let csrfToken: string | null = null;

const getCsrfToken = async (): Promise<string | null> => {
  if (csrfToken) return csrfToken;

  const res = await fetch(`${API_BASE_URL}/auth/csrf/restore`, {
    credentials: 'include',
  });

  if (!res.ok) throw new Error('Failed to fetch CSRF token');

  const data = await res.json();
  csrfToken = data.csrf_token;

  return csrfToken;
};

const makeRequest = async <T>(endpoint: string, method: string = 'GET', data?: unknown): Promise<T> => {
  const options: RequestInit = {
    method,
    credentials: 'include',
    headers: {
      'Content-Type': 'application/json',
    },
  };

  if (method !== 'GET') {
    const token = await getCsrfToken();
    if (token) {
      options.headers = {
        ...options.headers,
        'X-CSRFToken': token,
      };
    }
  }

  if (data) {
    options.body = JSON.stringify(data);
  }

  const response = await fetch(`${API_BASE_URL}${endpoint}`, options);

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw {
      status: response.status,
      data: errorData,
    };
  }

  if (response.status === 204) {
    return {} as T;
  }

  return response.json();
};

export const useNotes = () => {
  const queryClient = useQueryClient();
  const [currentNote, setCurrentNote] = useState<Note | null>(null);
  const [error, setError] = useState<string | null>(null);

  const {
    data: notes = [],
    isLoading: isLoadingNotes,
    refetch: refetchNotes,
  } = useQuery({
    queryKey: ['notes'],
    queryFn: () => makeRequest<Note[]>('/notes'),
    staleTime: 5 * 60 * 1000, // 5 minutes
  });

  const fetchNote = async (id: string) => {
    try {
      const note = await makeRequest<Note>(`/notes/${id}`);
      setCurrentNote(note);
      return note;
    } catch (error: any) {
      setError(error.data?.error || 'Failed to fetch note');
      throw error;
    }
  };

  const createNoteMutation = useMutation({
    mutationFn: (data: CreateNoteDto) => makeRequest<Note>('/notes', 'POST', data),
    onSuccess: newNote => {
      queryClient.setQueryData(['notes'], (old: Note[] = []) => [newNote, ...old]);
      setCurrentNote(newNote);
      setError(null);
    },
    onError: (err: any) => {
      setError(err.data?.error || 'Failed to create note');
    },
  });

  const updateNoteMutation = useMutation({
    mutationFn: ({ id, data }: { id: string; data: UpdateNoteDto }) => makeRequest<Note>(`/notes/${id}`, 'PUT', data),
    onSuccess: updatedNote => {
      queryClient.setQueryData(['notes'], (old: Note[] = []) =>
        old.map(note => (note.id === updatedNote.id ? updatedNote : note))
      );

      // Update current note if it's the one being edited
      if (currentNote?.id === updatedNote.id) {
        setCurrentNote(updatedNote);
      }

      setError(null);
    },
    onError: (err: any) => {
      setError(err.data?.error || 'Failed to update note');
    },
  });

  const deleteNoteMutation = useMutation({
    mutationFn: (id: string) => makeRequest<{ message: string }>(`/notes/${id}`, 'DELETE'),
    onSuccess: (_, id) => {
      queryClient.setQueryData(['notes'], (old: Note[] = []) => old.filter(note => note.id !== id));

      // Clear current note if it was deleted
      if (currentNote?.id === id) {
        setCurrentNote(null);
      }

      setError(null);
    },
    onError: (err: any) => {
      setError(err.data?.error || 'Failed to delete note');
    },
  });

  const createNote = async (data: CreateNoteDto) => {
    try {
      return await createNoteMutation.mutateAsync(data);
    } catch (error) {
      throw error;
    }
  };

  const updateNote = async (id: string, data: UpdateNoteDto) => {
    try {
      return await updateNoteMutation.mutateAsync({ id, data });
    } catch (error) {
      throw error;
    }
  };

  const deleteNote = async (id: string) => {
    try {
      await deleteNoteMutation.mutateAsync(id);
    } catch (error) {
      throw error;
    }
  };

  const clearError = () => setError(null);

  return {
    notes,
    currentNote,
    isLoading:
      isLoadingNotes || createNoteMutation.isPending || updateNoteMutation.isPending || deleteNoteMutation.isPending,
    error,
    setCurrentNote,
    clearError,
    fetchNote,
    createNote,
    updateNote,
    deleteNote,
    refetchNotes,
  };
};

export const useExperiments = () => {
  const queryClient = useQueryClient();
  const [currentExperiment, setCurrentExperiment] = useState<Experiment | null>(null);
  const [error, setError] = useState<string | null>(null);

  const {
    data: experiments = [],
    isLoading: isLoadingExperiments,
    refetch: refetchExperiments,
  } = useQuery({
    queryKey: ['experiments'],
    queryFn: () => makeRequest<Experiment[]>('/experiments'),
    staleTime: 5 * 60 * 1000, // 5 minutes
  });

  const fetchExperiment = async (id: string) => {
    try {
      const experiment = await makeRequest<Experiment>(`/experiments/${id}`);
      setCurrentExperiment(experiment);
      return experiment;
    } catch (error: any) {
      setError(error.data?.error || 'Failed to fetch experiment');
      throw error;
    }
  };

  const createExperimentMutation = useMutation({
    mutationFn: (data: CreateExperimentDto) => makeRequest<Experiment>('/experiments', 'POST', data),
    onSuccess: newExperiment => {
      queryClient.setQueryData(['experiments'], (old: Experiment[] = []) => [newExperiment, ...old]);
      setCurrentExperiment(newExperiment);
      setError(null);
    },
    onError: (err: any) => {
      setError(err.data?.error || 'Failed to create experiment');
    },
  });

  const updateExperimentMutation = useMutation({
    mutationFn: ({ id, data }: { id: string; data: UpdateExperimentDto }) =>
      makeRequest<Experiment>(`/experiments/${id}`, 'PUT', data),
    onSuccess: updatedExperiment => {
      queryClient.setQueryData(['experiments'], (old: Experiment[] = []) =>
        old.map(exp => (exp.id === updatedExperiment.id ? updatedExperiment : exp))
      );

      // Update current experiment if it's the one being edited
      if (currentExperiment?.id === updatedExperiment.id) {
        setCurrentExperiment(updatedExperiment);
      }

      setError(null);
    },
    onError: (err: any) => {
      setError(err.data?.error || 'Failed to update experiment');
    },
  });

  const deleteExperimentMutation = useMutation({
    mutationFn: (id: string) => makeRequest<{ message: string }>(`/experiments/${id}`, 'DELETE'),
    onSuccess: (_, id) => {
      queryClient.setQueryData(['experiments'], (old: Experiment[] = []) => old.filter(exp => exp.id !== id));

      // Clear current experiment if it was deleted
      if (currentExperiment?.id === id) {
        setCurrentExperiment(null);
      }

      setError(null);
    },
    onError: (err: any) => {
      setError(err.data?.error || 'Failed to delete experiment');
    },
  });

  const createExperiment = async (data: CreateExperimentDto) => {
    try {
      return await createExperimentMutation.mutateAsync(data);
    } catch (error) {
      throw error;
    }
  };

  const updateExperiment = async (id: string, data: UpdateExperimentDto) => {
    try {
      return await updateExperimentMutation.mutateAsync({ id, data });
    } catch (error) {
      throw error;
    }
  };

  const deleteExperiment = async (id: string) => {
    try {
      await deleteExperimentMutation.mutateAsync(id);
    } catch (error) {
      throw error;
    }
  };

  const clearError = () => setError(null);

  return {
    experiments,
    currentExperiment,
    isLoading:
      isLoadingExperiments ||
      createExperimentMutation.isPending ||
      updateExperimentMutation.isPending ||
      deleteExperimentMutation.isPending,
    error,
    setCurrentExperiment,
    clearError,
    fetchExperiment,
    createExperiment,
    updateExperiment,
    deleteExperiment,
    refetchExperiments,
  };
};
